osm_fudge/index/bk_tree.py

```diff
@@ -0,0 +1,54 @@
+'''An implementation of bk-tree
+
+More: https://en.wikipedia.org/wiki/BK-tree
+'''
+
+class BKTree:
+    '''A bk-tree implementation and defined user methods
+       for insertion and querying the string
+    '''
+
+    def __init__(self, distanceMetrics):
+        self.tree = {}
+        self.distanceMetrics = distanceMetrics
+
+    def insert(self, item):
+        self.__insert(item, self.tree)
+
+    def __insert(self, str1, subtree):
+        if not subtree.keys() or len(subtree.keys()) == 0:
+            self.tree = {str1: {}}
+            return
+
+        root = list(subtree.keys())[0]
+        dist = self.distanceMetrics(root, str1)
+
+        if dist in subtree[root]:
+            self.__insert(str1, subtree[root][dist])
+        else:
+            subtree[root][dist] = {str1: {}}
+
+    def lookup(self, item, max_distance=None):
+        return self.__lookup(item, self.tree, max_distance)
+
+    def __lookup(self, str1, subtree, tolerance):
+        root = list(subtree.keys())[0]
+
+        results = []
+        stack = []
+        stack.append(subtree)
+
+        while len(stack) > 0:
+            element = stack.pop()
+            for tree_string in element.keys():
+                dist = self.distanceMetrics(tree_string, str1)
+
+                if dist <= tolerance:
+                    results.append(tree_string)
+                    yield tree_string
+                    minimum = max([0, dist - tolerance])
+                    maximum = dist + tolerance
+
+                    for d in range(minimum, maximum + 1):
+                        if d in element[tree_string].keys():
+                            stack.append(element[tree_string][d])
```

Comment:
These three lines can be quite inefficient:
- We create a potentially large range and iterate over it
- Only then do we filter out elements we actually want

We could do two things here:
- Iterate over all sub-trees and descent only if their are within the range
- Or: store the sub-tree labels in a sorted list and do a first binary search to find all elements greater then the minimum, and from there do a second binary search to find all elements smaller then the maximum + 1. Maybe grab pen and paper and visualize some examples here.

The binary search approach requires us to keep the labels sorted. We could require the user e.g. to insert all elements and then call a `freeze` method which sorts all labels. Or insert sorts for every insertion (expensive). The binary search approach will also only more efficient then doing a linear search when your tree get's wide.

Reply:
I am a little confused here for the second option. Right now the tree looks like {label: distanceKey1: {label : ....}, distanceKey2: {label: .....}} . So do we sort the labels itself or the distance keys ?

If we sort distanceKeys in ascending order, then we can do binary search twice and grab the keys with our range of minimum and maximum + 1 and descent only for intersecting distance keys.

But I don't understand how will it work out if sort the labels instead of distancekeys. If labels are placed in sorted list, will it guarantee that the distance metrics with first label will always be less than distance metrics with last label? 🤔